File: Assets/ToLua/Lua/user/ui/hero/model/breakthrough_data.lua
local t = {}
t.data = {}
local item = {}
item.__index = item

local game_res_data = require('ui/game/model/game_res_data')

function item.New(table)
	local o = {}
	setmetatable(o,item)
	o.id = tonumber(table.id)
	o.layer = tonumber(table.layer)
	o.quality = tonumber(table.quality)
	o.levelMin = tonumber(table.lock_lev)
	o.levelMax = tonumber(table.lock_release)
  
  o.costGoldGameResData = nil
  if table.cost_1 ~= '0' then
    o.costGoldGameResData = game_res_data.NewByString(table.cost_1)
  end
  
  o.costItemGameResData = nil
  if table.cost_2 ~= '0' then
    o.costItemGameResData = game_res_data.NewByString(table.cost_2)
  end
	return o
end

function t.GetDataById(id)
	return t.data[id]
end

function t.GetBreakthroughDataListByQuality (quality)
  local breakthroughDataListOfQuality = {}
  for k, v in ipairs(t.data) do
    if v.quality == quality and v.layer > 1 then
      table.insert(breakthroughDataListOfQuality, v)
    end
  end
  return breakthroughDataListOfQuality
end

function t.GetNextBreakthroughDataByRoleInfo (roleInfo)
  for k, v in ipairs(t.data) do
    if v.quality == roleInfo.heroData.quality and v.layer == roleInfo.breakthroughLevel + 1 then
      return v
    end
  end
  return nil
end

local function Start()
	local origin = dofile('breakthrough')
	origin.ForEach(function(id,table)
		local newItem = item.New(table)
		t.data[id] = newItem
	end)
end

Start()
return t

File: Assets/Code/Logic/CommonGameData/BreakthroughData.cs
using UnityEngine;
using System.Collections.Generic;
using Common.Util;
using Common.ResMgr;
using Logic.Game.Model;
using Logic.Role.Model;

namespace Logic.CommonGameData
{
	public class BreakthroughData
	{
		private static Dictionary<int, BreakthroughData> _breakthroughDataDictionary;

        public static Dictionary<int, BreakthroughData> GetBreakthroughDatas()
        {
            if (_breakthroughDataDictionary == null)
            {
                _breakthroughDataDictionary = CSVUtil.Parse<int, BreakthroughData>(ResPath.GetConfigFilePath("breakthrough"), "id");
            }
            return _breakthroughDataDictionary;
        }

		public static Dictionary<int, BreakthroughData> BreakthroughDataDictionary
		{
			get
			{
				if (_breakthroughDataDictionary == null)
                {
                    GetBreakthroughDatas();
				}
				return _breakthroughDataDictionary;
			}
		}

		public static BreakthroughData GetRoleCurrentBreakthroughData (RoleInfo roleInfo)
		{
			BreakthroughData roleCurrentBreakthroughData = null;
			List<BreakthroughData> breakthroughDataList = BreakthroughDataDictionary.GetValues();
			for (int i = 0, count = breakthroughDataList.Count; i < count; i++)
			{
				if (breakthroughDataList[i].quality == (int)roleInfo.heroData.roleQuality
				    && breakthroughDataList[i].layer == roleInfo.breakthroughLevel)
				{
					roleCurrentBreakthroughData = breakthroughDataList[i];
				}
			}
			return roleCurrentBreakthroughData;
		}

		[CSVElement("id")]
		public int id;

		[CSVElement("layer")]
		public int layer;

		[CSVElement("quality")]
		public int quality;

		[CSVElement("lock_lev")]
		public int levelMin;

		[CSVElement("lock_release")]
		public int levelMax;

		public GameResData costGoldGameResData = null;
		[CSVElement("cost_1")]
		public string cost1
		{
			set
			{
				if (value.ToInt32() != 0)
				{
					costGoldGameResData = new GameResData(value);
				}
			}
		}

		public Logic.Game.Model.GameResData costItemGameResData = null;
		[CSVElement("cost_2")]
		public string cost2
		{
			set
			{
				if (value.ToInt32() != 0)
				{
					costItemGameResData = new GameResData(value);
				}
			}
		}
	}
}
